govern/data-security/krb-1.2.1/src/man/ktutil.man
.\" Man page generated from reStructuredText.
.
.TH "KTUTIL" "1" " " "1.20" "MIT Kerberos"
.SH NAME
ktutil \- Kerberos keytab file maintenance utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBktutil\fP
.SH DESCRIPTION
.sp
The ktutil command invokes a command interface from which an
administrator can read, write, or edit entries in a keytab.  (Kerberos
V4 srvtab files are no longer supported.)
.SH COMMANDS
.SS list
.INDENT 0.0
.INDENT 3.5
\fBlist\fP [\fB\-t\fP] [\fB\-k\fP] [\fB\-e\fP]
.UNINDENT
.UNINDENT
.sp
Displays the current keylist.  If \fB\-t\fP, \fB\-k\fP, and/or \fB\-e\fP are
specified, also display the timestamp, key contents, or enctype
(respectively).
.sp
Alias: \fBl\fP
.SS read_kt
.INDENT 0.0
.INDENT 3.5
\fBread_kt\fP \fIkeytab\fP
.UNINDENT
.UNINDENT
.sp
Read the Kerberos V5 keytab file \fIkeytab\fP into the current keylist.
.sp
Alias: \fBrkt\fP
.SS write_kt
.INDENT 0.0
.INDENT 3.5
\fBwrite_kt\fP \fIkeytab\fP
.UNINDENT
.UNINDENT
.sp
Write the current keylist into the Kerberos V5 keytab file \fIkeytab\fP\&.
.sp
Alias: \fBwkt\fP
.SS clear_list
.INDENT 0.0
.INDENT 3.5
\fBclear_list\fP
.UNINDENT
.UNINDENT
.sp
Clear the current keylist.
.sp
Alias: \fBclear\fP
.SS delete_entry
.INDENT 0.0
.INDENT 3.5
\fBdelete_entry\fP \fIslot\fP
.UNINDENT
.UNINDENT
.sp
Delete the entry in slot number \fIslot\fP from the current keylist.
.sp
Alias: \fBdelent\fP
.SS add_entry
.INDENT 0.0
.INDENT 3.5
\fBadd_entry\fP {\fB\-key\fP|\fB\-password\fP} \fB\-p\fP \fIprincipal\fP
\fB\-k\fP \fIkvno\fP [\fB\-e\fP \fIenctype\fP] [\fB\-f\fP|\fB\-s\fP \fIsalt\fP]
.UNINDENT
.UNINDENT
.sp
Add \fIprincipal\fP to keylist using key or password.  If the \fB\-f\fP flag
is specified, salt information will be fetched from the KDC; in this
case the \fB\-e\fP flag may be omitted, or it may be supplied to force a
particular enctype.  If the \fB\-f\fP flag is not specified, the \fB\-e\fP
flag must be specified, and the default salt will be used unless
overridden with the \fB\-s\fP option.
.sp
Alias: \fBaddent\fP
.SS list_requests
.INDENT 0.0
.INDENT 3.5
\fBlist_requests\fP
.UNINDENT
.UNINDENT
.sp
Displays a listing of available commands.
.sp
Aliases: \fBlr\fP, \fB?\fP
.SS quit
.INDENT 0.0
.INDENT 3.5
\fBquit\fP
.UNINDENT
.UNINDENT
.sp
Quits ktutil.
.sp
Aliases: \fBexit\fP, \fBq\fP
.SH EXAMPLE
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
ktutil:  add_entry \-password \-p alice@BLEEP.COM \-k 1 \-e
    aes128\-cts\-hmac\-sha1\-96
Password for alice@BLEEP.COM:
ktutil:  add_entry \-password \-p alice@BLEEP.COM \-k 1 \-e
    aes256\-cts\-hmac\-sha1\-96
Password for alice@BLEEP.COM:
ktutil:  write_kt alice.keytab
ktutil:
.ft P
.fi
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.SH ENVIRONMENT
.sp
See kerberos(7) for a description of Kerberos environment
variables.
.SH SEE ALSO
.sp
kadmin(1), kdb5_util(8), kerberos(7)
.SH AUTHOR
MIT
.SH COPYRIGHT
1985-2022, MIT
.\" Generated by docutils manpage writer.
.


kd/CHANGELOG.md
# Change Log


## [v0.4.0](https://github.com/koordinator-sh/koordinator/tree/v0.4.0) (2022-05-31)
[Full Changelog](https://github.com/koordinator-sh/koordinator/compare/v0.3.1...v0.4.0)

**Features and improvements:**

- Introduce koord-runtime-proxy [\#171](https://github.com/koordinator-sh/koordinator/pull/171)
- koord-runtime-proxy supports docker proxy [\#64](https://github.com/koordinator-sh/koordinator/issues/64)
- Support Memory eviction lower percent [\#85](https://github.com/koordinator-sh/koordinator/issues/85)
- Support load-aware scheduling [\#135](https://github.com/koordinator-sh/koordinator/pull/135)
- Support BE Pods eviction based on satisfaction [\#147](https://github.com/koordinator-sh/koordinator/issues/147)
- Support group identity [\#154](https://github.com/koordinator-sh/koordinator/issues/154)

**Fixed bugs:**

- Use the limit as request for BE Pods [\#49](https://github.com/koordinator-sh/koordinator/pull/129)

**Merged pull requests:**

- fix be container memory request [\#129](https://github.com/koordinator-sh/koordinator/pull/129) ([shinytang6](https://github.com/shinytang6))
- add koordlet runtime design [\#123](https://github.com/koordinator-sh/koordinator/pull/123) ([zwzhang0107](https://github.com/zwzhang0107))
- support memoryEvictLowerPercent in memory evict [\#132](https://github.com/koordinator-sh/koordinator/pull/132) ([shinytang6](https://github.com/shinytang6))
- add validation for CRD [\#133](https://github.com/koordinator-sh/koordinator/pull/133) ([jasonliu747](https://github.com/jasonliu747))
- Modify memqos wmark ratio doc desc [\#142](https://github.com/koordinator-sh/koordinator/pull/142) ([tianzichenone](https://github.com/tianzichenone))
- proposal load-aware scheduling plugin [\#135](https://github.com/koordinator-sh/koordinator/pull/135) ([eahydra](https://github.com/eahydra))
- koordlet: support NodeMetricCollectPolicy [\#157](https://github.com/koordinator-sh/koordinator/pull/157) ([eahydra](https://github.com/eahydra))
- koord-scheduler: Support load aware scheduling [\#159](https://github.com/koordinator-sh/koordinator/pull/159) ([eahydra](https://github.com/eahydra))
- koordlet: support collect BE CPU metric [\#158](https://github.com/koordinator-sh/koordinator/pull/158) ([jasonliu747](https://github.com/jasonliu747))
- apis: introduce cpu evict fields in NodeSLO [\#161](https://github.com/koordinator-sh/koordinator/pull/161) ([jasonliu747](https://github.com/jasonliu747))
- koordlet: support cpu evict feature [\#169](https://github.com/koordinator-sh/koordinator/pull/169) ([jasonliu747](https://github.com/jasonliu747))
- Add pod annotations/labels for container level hook [\#165](https://github.com/koordinator-sh/koordinator/pull/165) ([honpey](https://github.com/honpey))
- Introduce image service proxy under cri scenario [\#168](https://github.com/koordinator-sh/koordinator/pull/168) ([honpey](https://github.com/honpey))
- koord-runtime-proxy: refactor codes about store and resource-exectutor [\#170](https://github.com/koordinator-sh/koordinator/pull/170) ([honpey](https://github.com/honpey))
- Introduce main for koord-runtime-proxy [\#171](https://github.com/koordinator-sh/koordinator/pull/171) ([honpey](https://github.com/honpey))
- Add the koord-runtime-proxy design doc [\#178](https://github.com/koordinator-sh/koordinator/pull/178) ([honpey](https://github.com/honpey))
- koord-runtime-proxy supports docker proxy [\#128](https://github.com/koordinator-sh/koordinator/pull/128) ([ZYecho](https://github.com/ZYecho))
- add group identity plugin [\#166](https://github.com/koordinator-sh/koordinator/pull/166) ([zwzhang0107](https://github.com/zwzhang0107))
- use T.TempDir to create temporary test directory [\#151](https://github.com/koordinator-sh/koordinator/pull/151) ([Juneezee](https://github.com/Juneezee))
- update codecov configuration [\#131](https://github.com/koordinator-sh/koordinator/pull/131) ([saintube](https://github.com/saintube))

**New Contributors**

- [shinytang6](https://github.com/shinytang6) made their contributions in [\#129](https://github.com/koordinator-sh/koordinator/pull/129), [\#132](https://github.com/koordinator-sh/koordinator/pull/132)
- [tianzichenone](https://github.com/tianzichenone) made their first contribution in [\#142](https://github.com/koordinator-sh/koordinator/pull/142)
- [Juneezee](https://github.com/Juneezee) made their first contribution in [\#151](https://github.com/koordinator-sh/koordinator/pull/151)
- [ZYecho](https://github.com/ZYecho) made their first contribution in [\#128](https://github.com/koordinator-sh/koordinator/pull/128)


## [v0.3.0](https://github.com/koordinator-sh/koordinator/tree/v0.3.0) (2022-05-07)
[Full Changelog](https://github.com/koordinator-sh/koordinator/compare/v0.2.0...v0.3.0)

**Features and improvements:**

- Support CPU burst strategy [\#52](https://github.com/koordinator-sh/koordinator/issues/52)
- Support Memory QoS strategy [\#55](https://github.com/koordinator-sh/koordinator/issues/55)
- Support LLC and MBA isolation strategy [\#56](https://github.com/koordinator-sh/koordinator/issues/56)
- Protocol design between runtime-manager and hook server [\#62](https://github.com/koordinator-sh/koordinator/issues/62)
- Improve overall code coverage from 39% to 56% [\#69](https://github.com/koordinator-sh/koordinator/issues/69)

**Fixed bugs:**

- when deploy on ACK 1.18.1 koord-manager pod always crash [\#49](https://github.com/koordinator-sh/koordinator/issues/49)
- Handle unexpected CPU info in case of koordlet panic [\#90](https://github.com/koordinator-sh/koordinator/issues/90)

**Merged pull requests:**

- New feature: cpu burst strategy [\#73](https://github.com/koordinator-sh/koordinator/pull/73) ([stormgbs](https://github.com/stormgbs))
- Introduce protocol between RuntimeManager and RuntimeHookServer [\#76](https://github.com/koordinator-sh/koordinator/pull/76) ([honpey](https://github.com/honpey))

govern/data-security/krb-1.2.1/src/kadmin/ktutil/ktutil.c
/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
/* kadmin/ktutil/ktutil.c - SS user interface for ktutil */
/*
 * Copyright 1995, 1996, 2008 by the Massachusetts Institute of Technology.
 * All Rights Reserved.
 *
 * Export of this software from the United States of America may
 *   require a specific license from the United States Government.
 *   It is the responsibility of any person or organization contemplating
 *   export to obtain such a license before exporting.
 *
 * WITHIN THAT CONSTRAINT, permission to use, copy, modify, and
 * distribute this software and its documentation for any purpose and
 * without fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright notice and
 * this permission notice appear in supporting documentation, and that
 * the name of M.I.T. not be used in advertising or publicity pertaining
 * to distribution of the software without specific, written prior
 * permission.  Furthermore if you modify this software you must label
 * your software as modified software and not distribute it in such a
 * fashion that it might be confused with the original M.I.T. software.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is" without express
 * or implied warranty.
 */

#include "k5-int.h"
#include "ktutil.h"
#include <com_err.h>
#include <locale.h>
#include "adm_proto.h"
#include <ss/ss.h>
#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

extern ss_request_table ktutil_cmds;
krb5_context kcontext;
krb5_kt_list ktlist = NULL;

int main(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;
    int sci_idx;

    setlocale(LC_ALL, "");
    retval = krb5_init_context(&kcontext);
    if (retval) {
        com_err(argv[0], retval, _("while initializing krb5"));
        exit(1);
    }
    sci_idx = ss_create_invocation("ktutil", "5.0", (char *)NULL,
                                   &ktutil_cmds, &retval);
    if (retval) {
        ss_perror(sci_idx, retval, _("creating invocation"));
        exit(1);
    }
    retval = ss_listen(sci_idx);
    ktutil_free_kt_list(kcontext, ktlist);
    exit(0);
}

void ktutil_clear_list(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;

    if (argc != 1) {
        fprintf(stderr, _("%s: invalid arguments\n"), argv[0]);
        return;
    }
    retval = ktutil_free_kt_list(kcontext, ktlist);
    if (retval)
        com_err(argv[0], retval, _("while freeing ktlist"));
    ktlist = NULL;
}

void ktutil_read_v5(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;

    if (argc != 2) {
        fprintf(stderr, _("%s: must specify keytab to read\n"), argv[0]);
        return;
    }
    retval = ktutil_read_keytab(kcontext, argv[1], &ktlist);
    if (retval)
        com_err(argv[0], retval, _("while reading keytab \"%s\""), argv[1]);
}

void ktutil_read_v4(argc, argv)
    int argc;
    char *argv[];
{
    fprintf(stderr, _("%s: reading srvtabs is no longer supported\n"),
            argv[0]);
}

void ktutil_write_v5(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;

    if (argc != 2) {
        fprintf(stderr, _("%s: must specify keytab to write\n"), argv[0]);
        return;
    }
    retval = ktutil_write_keytab(kcontext, ktlist, argv[1]);
    if (retval)
        com_err(argv[0], retval, _("while writing keytab \"%s\""), argv[1]);
}

void ktutil_write_v4(argc, argv)
    int argc;
    char *argv[];
{
    fprintf(stderr, _("%s: writing srvtabs is no longer supported\n"),
            argv[0]);
}

void ktutil_add_entry(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;
    char *princ = NULL;
    char *enctype = NULL;
    krb5_kvno kvno = 0;
    int use_pass = 0, use_key = 0, use_kvno = 0, fetch = 0, i;
    char *salt = NULL;

    for (i = 1; i < argc; i++) {
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-p", 2)) {
            princ = argv[++i];
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-k", 2)) {
            kvno = (krb5_kvno) atoi(argv[++i]);
            use_kvno++;
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-e", 2)) {
            enctype = argv[++i];
            continue;
        }
        if ((strlen(argv[i]) == 9) && !strncmp(argv[i], "-password", 9)) {
            use_pass++;
            continue;
        }
        if ((strlen(argv[i]) == 4) && !strncmp(argv[i], "-key", 4)) {
            use_key++;
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-s", 2)) {
            salt = argv[++i];
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-f", 2))
            fetch++;
    }

    if (princ == NULL || use_pass + use_key != 1 || !use_kvno ||
        (fetch && salt != NULL)) {
        fprintf(stderr, _("usage: %s (-key | -password) -p principal "
                          "-k kvno [-e enctype] [-f|-s salt]\n"), argv[0]);
        return;
    }
    if (!fetch && enctype == NULL) {
        fprintf(stderr, _("enctype must be specified if not using -f\n"));
        return;
    }

    retval = ktutil_add(kcontext, &ktlist, princ, fetch, kvno, enctype,
                        use_pass, salt);
    if (retval)
        com_err(argv[0], retval, _("while adding new entry"));
}

void ktutil_delete_entry(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;

    if (argc != 2) {
        fprintf(stderr, _("%s: must specify entry to delete\n"), argv[0]);
        return;
    }
    retval = ktutil_delete(kcontext, &ktlist, atoi(argv[1]));
    if (retval)
        com_err(argv[0], retval, _("while deleting entry %d"), atoi(argv[1]));
}

void ktutil_list(argc, argv)
    int argc;
    char *argv[];
{
    krb5_error_code retval;
    krb5_kt_list lp;
    int show_time = 0, show_keys = 0, show_enctype = 0;
    int i;
    unsigned int j;
    char *pname;

    for (i = 1; i < argc; i++) {
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-t", 2)) {
            show_time++;
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-k", 2)) {
            show_keys++;
            continue;
        }
        if ((strlen(argv[i]) == 2) && !strncmp(argv[i], "-e", 2)) {
            show_enctype++;
            continue;
        }

        fprintf(stderr, _("%s: usage: %s [-t] [-k] [-e]\n"), argv[0], argv[0]);
        return;
    }
    /* XXX Translating would disturb table alignment; skip for now. */
    if (show_time) {
        printf("slot KVNO Timestamp         Principal\n");
        printf("---- ---- ----------------- ---------------------------------------------------\n");
    } else {
        printf("slot KVNO Principal\n");
        printf("---- ---- ---------------------------------------------------------------------\n");
    }
    for (i = 1, lp = ktlist; lp; i++, lp = lp->next) {
        retval = krb5_unparse_name(kcontext, lp->entry->principal, &pname);
        if (retval) {
            com_err(argv[0], retval, "while unparsing principal name");
            return;
        }
        printf("%4d %4d ", i, lp->entry->vno);
        if (show_time) {
            char fmtbuf[18];
            char fill;
            time_t tstamp;

            tstamp = lp->entry->timestamp;
            lp->entry->timestamp = tstamp;
            fill = ' ';
            if (!krb5_timestamp_to_sfstring((krb5_timestamp)lp->entry->
                                            timestamp,
                                            fmtbuf,
                                            sizeof(fmtbuf),
                                            &fill))
                printf("%s ", fmtbuf);
        }
        printf("%40s", pname);
        if (show_enctype) {
            static char buf[256];
            if ((retval = krb5_enctype_to_name(lp->entry->key.enctype, FALSE,
                                               buf, sizeof(buf)))) {
                com_err(argv[0], retval,
                        _("While converting enctype to string"));
                return;
            }
            printf(" (%s) ", buf);
        }

        if (show_keys) {
            printf(" (0x");
            for (j = 0; j < lp->entry->key.length; j++)
                printf("%02x", lp->entry->key.contents[j]);
            printf(")");
        }
        printf("\n");
        free(pname);
    }
}
